core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -17,27 +17,78 @@
 
 package kafka.cluster
 
-import kafka.log.{Log, LogOffsetSnapshot}
+import kafka.log.{Log}
 import kafka.utils.Logging
-import kafka.server.{LogOffsetMetadata, LogReadResult, OffsetAndEpoch}
+import kafka.server.{LogOffsetMetadata, LogReadResult}
 import org.apache.kafka.common.{KafkaException, TopicPartition}
-import org.apache.kafka.common.errors.OffsetOutOfRangeException
-import org.apache.kafka.common.utils.Time
-
-class Replica(val brokerId: Int,
-              val topicPartition: TopicPartition,
-              time: Time = Time.SYSTEM,
-              initialHighWatermarkValue: Long = 0L,
-              @volatile var log: Option[Log] = None) extends Logging {
-  // the high watermark offset value, in non-leader replicas only its message offsets are kept
-  @volatile private[this] var highWatermarkMetadata = new LogOffsetMetadata(initialHighWatermarkValue)
+
+abstract class Replica(val brokerId: Int,
+                       val topicPartition: TopicPartition) extends Logging {
+
+  // lastCaughtUpTimeMs is the largest time t such that the offset of most recent FetchRequest from this follower >=
+  // the LEO of leader at time t. This is used to determine the lag of this follower and ISR of this partition.
+  @volatile private[this] var _lastCaughtUpTimeMs = 0L
+
+  def logStartOffset: Long
+  def logEndOffsetMetadata: LogOffsetMetadata
+
+  def logEndOffset: Long = logEndOffsetMetadata.messageOffset
+
+  def lastCaughtUpTimeMs: Long = _lastCaughtUpTimeMs
+
+  def lastCaughtUpTimeMs_=(lastCaughtUpTimeMs: Long) {
+    _lastCaughtUpTimeMs = lastCaughtUpTimeMs
+    trace(s"Setting log caught offset time for replica $brokerId, partition $topicPartition to $lastCaughtUpTimeMs")
+  }
+
+  def resetLastCaughtUpTime(curLeaderLogEndOffset: Long, curTimeMs: Long, lastCaughtUpTimeMs: Long): Unit = {
+    throw new KafkaException(
+      s"Method not implemented for partition $topicPartition and broker id $brokerId")
+  }
+
+  def updateLogReadResult(logReadResult: LogReadResult): Unit = {
+    throw new KafkaException(
+      s"Method not implemented for partition $topicPartition and broker id $brokerId")
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case other: Replica => brokerId == other.brokerId && topicPartition == other.topicPartition
+    case _ => false
+  }
+
+  override def hashCode: Int = 31 + topicPartition.hashCode + 17 * brokerId
+}
+
+class LocalReplica(brokerId: Int,
+                   topicPartition: TopicPartition,
+                   @volatile var log: Log) extends Replica(brokerId, topicPartition) {
```

Comment:
Should we have this here given that we are now storing it as part `Partition`?

By having this here don't we have the requirement that both variable to be updated atomically?

Reply:
Good point. Currently this variable and the "Partition::log" variable are created in Partition::getOrCreateReplica and updated in Partition::maybeReplaceCurrentWithFutureReplica. While we can say that create happens only once (probably when Partition is getting initialized) and futureReplica to localReplica switch happens in a lock, it doesn't solve the atomicity problem.

One solution is to write a wapper and use that, Replica used to be that wrapper :-)  I added a LogInfo wrapper in my first version but that felt a bit kludgy. I will think more about it, but any other suggestion is welcome!!

Reply:
One possible solution is what @hachikuji suggested with making `Replica` (or possible a different name) a `trait` and having both `Log` and `RemoteReplica` implement it.

Another option if we don't have to make `Replica` part of the type hierarchy for `Log` is extension methods: https://docs.scala-lang.org/overviews/core/value-classes.html. We do this here: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/api/package.scala#L39

Reply:
I was looking through uses of `allReplicasMap`. I suspect that after this refactor, we may not have a strong reason for a shared trait for both local and remote replicas. Basically `RemoteReplica` can be just `Replica`.

Reply:
Done. No more LocalReplica and multiple copies of log.